### PR TITLE
Enhance flow execution

### PR DESCRIPTION
**Pull request by adracus**
**What this PR does / why we need it**:
Enhance flow creation / execution / monitoring

```improvement operator
NONE
```

### CODE REVIEW

File: pkg/controller/shoot/shoot_control_reconcile.go [hunk 1]
```diff
@@ -52,58 +52,226 @@ func (c *defaultControl) reconcileShoot(o *operation.Operation, operationType ga
...
+		defaultInterval = 5 * time.Second
+		managedDNS      = o.Shoot.Info.Spec.DNS.Provider != gardenv1beta1.DNSUnmanaged
+		isCloud         = o.Shoot.Info.Spec.Cloud.Local == nil
+
+		g = flow.NewGraph("Shoot cluster reconciliation")
+
+		deployNamespace = g.Add(flow.Task{
+			Name: "Deploy Namespace",
+			Fn: flow.TaskFn(botanist.DeployNamespace).
+				RetryUntilTimeout(defaultInterval, defaultTimeout),
```
Comment (rfranzke) [inline pkg/controller/shoot/shoot_control_reconcile.go]: Let's put that one line above to improve readability (same for the following tasks).

File: pkg/operation/hybridbotanist/machines.go [hunk 2]
```diff
@@ -104,7 +104,7 @@ func (b *HybridBotanist) ReconcileMachines() error {
 
 	// Wait until all generated machine deployments are healthy/available.
 	if err := b.waitUntilMachineDeploymentsAvailable(wantedMachineDeployments); err != nil {
-		return common.DetermineErrorCode(fmt.Sprintf("Failed while waiting for all machine deployments to be ready: '%s'", err.Error()))
+		return common.NewTerraformError(fmt.Sprintf("Failed while waiting for all machine deployments to be ready: '%s'", err.Error()))
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/machines.go]: Let's not call the function `NewTerraformError` as it is also used here where Terraform does not play any role at all.

File: pkg/operation/hybridbotanist/machines.go [hunk 3]
```diff
@@ -171,7 +171,7 @@ func (b *HybridBotanist) DestroyMachines() error {
 
 	// Wait until all machine resources have been properly deleted.
 	if err := b.waitUntilMachineResourcesDeleted(machineClassPlural); err != nil {
-		return common.DetermineErrorCode(fmt.Sprintf("Failed while waiting for all machine resources to be deleted: '%s'", err.Error()))
+		return common.NewTerraformError(fmt.Sprintf("Failed while waiting for all machine resources to be deleted: '%s'", err.Error()))
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/machines.go]: Let's not call the function `NewTerraformError` as it is also used here where Terraform does not play any role at all.

**Review by rfranzke — CHANGES_REQUESTED** [review]
1. When I try to simulate an error, e.g. in the `*Botanist.DeployNamespace()` function like this:

```golang
func (b *Botanist) DeployNamespace() error {
	return fmt.Errorf("couldn't deploy ns")
}
```

then I see log messages like this:
```
time="2018-09-04T09:31:59+02:00" level=error msg="encountered error while task Deploy Namespace: task \"Deploy Namespace\" failed: timed out waiting for the condition" flow="Shoot cluster reconciliation" opid=Q93vEcHn shoot=garden-dev/rfranzke-a1
time="2018-09-04T09:31:59+02:00" level=info msg="Task %!s(MISSING) failed" flow="Shoot cluster reconciliation" opid=Q93vEcHn shoot=garden-dev/rfranzke-a1
...
time="2018-09-04T09:31:59+02:00" level=error msg="Failed to reconcile shoot \"rfranzke-a1\": 1 error occurred:\n\t* task \"Deploy Namespace\" failed: timed out waiting for the condition\n\n\nflow \"Shoot cluster reconciliation\" failed\ngithub.com/gardener/gardener/pkg/utils/flow.(*execution).run\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/utils/flow/flow.go:238\ngithub.com/gardener/gardener/pkg/utils/flow.(*Flow).Run\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/utils/flow/flow.go:87\ngithub.com/gardener/gardener/pkg/controller/shoot.(*defaultControl).reconcileShoot\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/shoot/shoot_control_reconcile.go:249\ngithub.com/gardener/gardener/pkg/controller/shoot.(*defaultControl).ReconcileShoot\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/shoot/shoot_control.go:250\ngithub.com/gardener/gardener/pkg/controller/shoot.(*Controller).reconcileShootKey\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/shoot/shoot_control.go:114\ngithub.com/gardener/gardener/pkg/controller/shoot.(*Controller).(github.com/gardener/gardener/pkg/controller/shoot.reconcileShootKey)-fm\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/shoot/shoot.go:225\ngithub.com/gardener/gardener/pkg/controller/utils.worker.func1.1\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/utils/worker.go:52\ngithub.com/gardener/gardener/pkg/controller/utils.worker.func1\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/utils/worker.go:61\ngithub.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88\ngithub.com/gardener/gardener/pkg/controller/utils.CreateWorker.func1\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/utils/worker.go:33\nruntime.goexit\n\t/usr/local/Cellar/go/1.10.2/libexec/src/runtime/asm_amd64.s:2361\nerror during Shoot reconcile flow\ngithub.com/gardener/gardener/pkg/controller/shoot.(*defaultControl).reconcileShoot\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/shoot/shoot_control_reconcile.go:252\ngithub.com/gardener/gardener/pkg/controller/shoot.(*defaultControl).ReconcileShoot\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/shoot/shoot_control.go:250\ngithub.com/gardener/gardener/pkg/controller/shoot.(*Controller).reconcileShootKey\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/shoot/shoot_control.go:114\ngithub.com/gardener/gardener/pkg/controller/shoot.(*Controller).(github.com/gardener/gardener/pkg/controller/shoot.reconcileShootKey)-fm\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/shoot/shoot.go:225\ngithub.com/gardener/gardener/pkg/controller/utils.worker.func1.1\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/utils/worker.go:52\ngithub.com/gardener/gardener/pkg/controller/utils.worker.func1\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/utils/worker.go:61\ngithub.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88\ngithub.com/gardener/gardener/pkg/controller/utils.CreateWorker.func1\n\t/Users/rfranzke/go/src/github.com/gardener/gardener/pkg/controller/utils/worker.go:33\nruntime.goexit\n\t/usr/local/Cellar/go/1.10.2/libexec/src/runtime/asm_amd64.s:2361" opid=Q93vEcHn shoot=garden-dev/rfranzke-a1
```
The actual error message ("couldn't deploy ns") is hidden.
Also, couldn't we merge the first two log messages? Or, just remove the second one (the error is already printed before).


1. I see logs like this (missing task name):
```
time="2018-09-04T09:32:30+02:00" level=info msg="Task %!s(MISSING) succeeded" flow="Shoot cluster reconciliation" opid=K6QreUWc shoot=garden-dev/rfranzke-a1
```

1. The `.status.lastError` contains:

```yaml
status:
  lastError:
    description: "error during Shoot reconcile flow: flow \"Shoot cluster reconciliation\"
      failed: 1 error occurred:\n\t* task \"Deploy Namespace\" failed: timed out
      waiting for the condition\n\n"
```

Could we enhance this message (as it will be seen by end-users)? Something like: `Flow "Shoot cluster reconciliation" failed due to 1 error:\n\t*Task "Deploy Namespace" failed: <error>\n\n` or `Flow "Shoot cluster reconciliation" successfully executed.

1. Now that we have more descriptive ways for the individual flow tasks, the message in the `.status.lastOperation` during a processing operation could be enhanced. Currently, something like this is shown:
```yaml
status:
  lastOperation:
    description: Executing Deploy Namespace, Deploy Kube API Server Service.
```
Maybe we can remove the "Executing" and find better readable names for the tasks, e.g. `Deploying namespace`, `Deploying load balancer service for kube-apiserver`?

**Comment by adracus** [comment]
Adapted code, verified locally

**Review by rfranzke — APPROVED** [review]

**Pull request merged**